File: src/bin/day10.rs
//! Day 10: Pipe Maze
//!
//! <https://adventofcode.com/2023/day/10>
use std::{
    collections::{
        VecDeque,
        HashSet,
        HashMap,
    },
    fmt::Display,
};
use aoc_2023::Solution;

static GO_LEFT_PIPES: [u8; 3] =
    [b'-', b'J', b'7'];
static GO_RIGHT_PIPES: [u8; 3] =
    [b'-', b'L', b'F'];
static GO_UP_PIPES: [u8; 3] =
    [b'|', b'J', b'L'];
static GO_DOWN_PIPES: [u8; 3] =
    [b'|', b'7', b'F'];

pub struct Day10;

impl Day10 {
    #[inline]
    fn get_grid<T: Display>(inp: T) -> Vec<Vec<u8>> {
        inp
            .to_string()
            .lines()
            .map(|row| row
                .as_bytes()
                .to_vec()
            )
            .collect::<Vec<Vec<u8>>>()
    }

    fn get_starting_pos(grid: &[Vec<u8>]) -> (usize, usize) {
        for (i, row) in grid
            .iter()
            .enumerate()
        {
            if let Some(j) = row
                .iter()
                .position(|c| *c == b'S')
            {
                return (i, j)
            }
        }
        panic!("No 'S' character found in grid")
    }

    fn get_loop(grid: &[Vec<u8>]) -> HashSet<(usize, usize)> {
        let starting_coords = Self::get_starting_pos(grid);

        let mut nodes = HashSet::from([starting_coords]);
        let mut to_check = VecDeque::from([starting_coords]);

        while let Some((curr_row, curr_col)) = to_check.pop_front() {
            let curr_tile = grid[curr_row][curr_col];

            let neighbors = [
                (curr_row.wrapping_sub(1), curr_col,
                    GO_UP_PIPES,
                    GO_DOWN_PIPES,
                ),
                (curr_row + 1, curr_col,
                    GO_DOWN_PIPES,
                    GO_UP_PIPES,
                ),
                (curr_row, curr_col.wrapping_sub(1),
                    GO_LEFT_PIPES,
                    GO_RIGHT_PIPES,
                ),
                (curr_row, curr_col + 1,
                    GO_RIGHT_PIPES,
                    GO_LEFT_PIPES,
                ),
            ];
            for (next_row, next_col, pipes, co_pipes) in neighbors {
                if let Some(next_tile) = grid
                    .get(next_row)
                    .and_then(|row| row.get(next_col))
                {
                    let next_coord = (next_row, next_col);
                    if (pipes.contains(&curr_tile)
                        || curr_tile == b'S')
                        && co_pipes.contains(next_tile)
                        && !nodes.contains(&next_coord)
                    {
                        nodes.insert(next_coord);
                        to_check.push_back(next_coord);
                    }
                }
            }
        }
        nodes
    }

    /// Reformats the grid using unicode characters to help better visualize the pipes
    ///
    /// All pipes that are not part of the loop are replaced with a "."
    pub fn display_grid<T: Display>(&self, inp: T) -> String {
        let mapping = HashMap::from([
            (b'|', '│'),
            (b'-', '─'),
            (b'J', '┘'),
            (b'7', '┐'),
            (b'L', '└'),
            (b'F', '┌'),
        ]);
        let grid = Self::get_grid(inp);
        let nodes = Self::get_loop(&grid);

        let formatted = grid
            .iter()
            .enumerate()
            .map(|(i, row)| format!(
                "{i:>3}| {}",
                row
                    .iter()
                    .enumerate()
                    .map(|(j, tile)|
                        if nodes.contains(&(i, j)) {
                            *mapping.get(tile)
                                .unwrap_or(&(*tile as char))
                        }
                        else { '.' }
                        .to_string()
                    )
                    .collect::<String>()
                )
            )
            .collect::<Vec<String>>()
            .join("\n");

        format!("\n{formatted}\n")
    }
}

impl Solution for Day10 {
    const NAME: &'static str = "Pipe Maze";

    fn part_one<T: Display>(&self, inp: T) -> usize {
        Self::get_loop(
            &Self::get_grid(inp)
        )
        .len() / 2
    }

    fn part_two<T: Display>(&self, inp: T) -> usize {
        let grid = Self::get_grid(inp);
        let nodes = Self::get_loop(&grid);

        let mut area = 0;
        for (i, row) in grid
            .into_iter()
            .enumerate()
        {
            let mut downwards = 0;
            let mut upwards = 0;

            for (j, tile) in row
                .into_iter()
                .enumerate()
            {
                if nodes.contains(&(i, j)) {
                    if GO_DOWN_PIPES.contains(&tile) {
                        downwards += 1;
                    }
                    if GO_UP_PIPES.contains(&tile) {
                        upwards += 1;
                    }
                } else if downwards % 2 == 1 && upwards % 2 == 1 {
                    area += 1;
                }
            }
        }
        area
    }

    fn run(&self, inp: String) {
        println!("{}", self.display_grid(&inp));

        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 7063);
        assert_eq!(p2, 589);
    }
}

fn main() {
    aoc_2023::run_day(10, &Day10);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}

File: aoc-py/solutions/day10.py
"""
Day 10: Pipe Maze

https://adventofcode.com/2023/day/10
"""
__all__ = ('Day10',)

from typing import ClassVar

from ..solution import Solution

class Day10(Solution):
    NAME: ClassVar[str] = 'Pipe Maze'

    GO_LEFT_PIPES: ClassVar[set[str]] = {'-', 'J', '7'}
    GO_RIGHT_PIPES: ClassVar[set[str]] = {'-', 'L', 'F'}
    GO_UP_PIPES: ClassVar[set[str]] = {'|', 'J', 'L'}
    GO_DOWN_PIPES: ClassVar[set[str]] = {'|', '7', 'F'}

    DISPLAY_CHAR_MAPPING: ClassVar[dict[str, str]] = {
        '|': '│',
        '-': '─',
        'J': '┘',
        '7': '┐',
        'L': '└',
        'F': '┌',
    }

    def _get_starting_pos(self, grid: list[str]) -> tuple[int, int]:
        for i, row in enumerate(grid):
            for j, char in enumerate(row):
                if char == 'S':
                    return i, j
        raise ValueError("No 'S' character found in grid")

    def _get_loop(self, grid: list[str]) -> set[tuple[int, int]]:
        starting_coords = self._get_starting_pos(grid)

        n_rows = len(grid)
        n_cols = len(grid[0])

        loop = {starting_coords}
        to_check = [starting_coords]
        while to_check:
            curr_row, curr_col = to_check.pop(0)
            curr_tile = grid[curr_row][curr_col]

            neighbors = (
                (curr_row - 1, curr_col, # check 1 tile up
                    self.GO_UP_PIPES,
                    self.GO_DOWN_PIPES
                ),
                (curr_row + 1, curr_col, # check 1 tile down
                    self.GO_DOWN_PIPES,
                    self.GO_UP_PIPES
                ),
                (curr_row, curr_col - 1, # check 1 tile left
                    self.GO_LEFT_PIPES,
                    self.GO_RIGHT_PIPES
                ),
                (curr_row, curr_col + 1, # check 1 tile right
                    self.GO_RIGHT_PIPES,
                    self.GO_LEFT_PIPES
                ),
            )
            for next_row, next_col, pipes, co_pipes in neighbors:
                if (
                    # ensure the row of the next tile doesn't overflow
                    next_row in range(n_rows)
                    # ensure the column of the next tile doesn't overflow
                    and next_col in range(n_cols)
                    # ensure the current tile is able to go in the direction (`pipes`)
                    and curr_tile in pipes | {'S'}
                    # ensure the next tile is able to accept the direction (`co_pipes`)
                    # `co_pipes` are the opposite direction of `pipes`
                    # i.e. a pipe going up needs the next tile to have a pipe that goes down
                    and grid[next_row][next_col] in co_pipes
                    # ensure the tile has not been traversed yet i.e. we are going backwards in the loop
                    and (next_coord := (next_row, next_col)) not in loop
                ):
                    to_check.append(next_coord)
                    loop.add(next_coord)
        return loop

    def part_one(self, inp: str) -> int:
        grid = inp.splitlines()
        loop = self._get_loop(grid)
        # the halfway mark of the loop is the furthest point from the start
        return len(loop) // 2

    def part_two(self, inp: str) -> int:
        """Solution to Part 2 using Ray Casting

        <https://en.wikipedia.org/wiki/Point_in_polygon>
        """
        grid = inp.splitlines()
        loop = self._get_loop(grid)

        area = 0
        for i, row in enumerate(grid):
            downwards = 0
            upwards = 0
            for j, char in enumerate(row):
                if (i, j) in loop:
                    if char in self.GO_DOWN_PIPES: # | ┌ ┐, downward facing pipes
                        downwards += 1
                    if char in self.GO_UP_PIPES:   # | └ ┘, upward facing pipes
                        upwards += 1
                elif downwards % 2 == 1 and upwards % 2 == 1:
                    # we hit a possible candidate for a tile that can be in or out the loop
                    # since we have been keeping track of the amount of "walls" we have hit
                    # we check if the amount of downward pointing walls and upward pointing walls are all even
                    # if so we can say that this tile is inside the loop
                    area += 1
        return area

    def display_grid(self, inp: str) -> str:
        """Reformats the grid using unicode characters to help better visualize the pipes

        All pipes that are not part of the loop are replaced with a "."
        """
        grid = inp.splitlines()
        loop = self._get_loop(grid)
        grid = [list(row) for row in grid]

        for i, row in enumerate(grid):
            for j, char in enumerate(row):
                if (i, j) in loop:
                    grid[i][j] = self.DISPLAY_CHAR_MAPPING.get(char, char)
                else:
                    grid[i][j] = '.'

        return '\n' + '\n'.join(
            f"{i:>3}| {''.join(row)}"
            for i, row in enumerate(grid, 1)
        ) + '\n'

    def run(self, inp: str) -> None:
        print(self.display_grid(inp))

        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 7063
        assert p2 == 589